doc/src/main/tut/structure.md

```diff
@@ -0,0 +1,222 @@
+    output: Path,
+)
+
+val configOpts = 
+  (remoteOpts.orNone, fileOpt.orNone, outputOpt)
+    .mapN(Config.apply)
+    .validate("must provide either uri or file")(c => c.remote.isDefined ^ c.file.isDefined)
+```
+
+With this change,
+it's no longer possible to define a `Config` that has a `timeout` but not a `uri`.
+This means we can get rid of one of our explicit validations;
+`decline` will do the equivalent input check automatically,
+and its autogenerated `Usage: ` info in the help output
+will also reflect the new structure.
+
+Other benefits include:
+- A single case class with twenty fields can get pretty unwieldy;
+  it becomes easy to accidentally pass arguments in the wrong order, for example.
+  Grouping arguments makes these errors less likely,
+  both because small groups are easier to see at a glance
+  and because the more explicit and specific your types are 
+  the more the compiler can help.
+- It can make it easier to validate early --
+  if we decided we wanted to ban custom timeouts
+  when the URI was `localhost`, for example,
+  we could add that validation to `remoteOpts` instead of `configOpts`.
+
+## Mutual exclusion
+
+`orElse` is often used for subcommands,
+but it works just as well for ordinary options.
+Since our input file is always either local or remote,
+using `orElse` can let us express that more directly.
+
+```scala mdoc:nest:to-string
+// Either would for two mutually-exclusive possibilities,
+// but a sealed trait is a bit more general.
+sealed trait InputConfig
+case class RemoteConfig(uri: URI, timeout: Duration) extends InputConfig
+case class LocalConfig(file: Path) extends InputConfig
+
+val remoteOpts = (uriOpt, timeoutOpt).mapN(RemoteConfig.apply)
+val localOpts = fileOpt.map(LocalConfig.apply)
+val inputOpts = remoteOpts orElse localOpts
+
+case class Config(
+    input: InputConfig,
+    queries: Path,
+)
+
+val configOpts = (inputOpts, outputOpt).mapN(Config.apply)
+```
+
+That gets rid of our final config validation;
+`decline` ensures that the user passes at least one of `--uri` or `--file`,
+but never both.
+
+It's often possible to replace ad-hoc validation with better data modelling like this,
+and it's almost always a good idea when you can:
+it simplifies the code,
+improves the error messages and usage texts that `decline` generates,
+and helps "make illegal states unrepresentable" in your program.
+
+## Config and effect style
+
+Through every step of the refactoring above,
+the shape of our command-line parser closely matches the `Config` datastructure;
+we build parsers for case classes using `mapN` and sealed traits using `orElse`,
+working our way up from simpler types to the full `Config`.
+(And once that's done,
+we hand the whole thing off to some `run` function
+that does the actual work.)
+This "config" pattern is a very common way to structure an application using `decline`,
+and it's easy to test:
+you might write unit tests that pass different arguments to the parser
```

Comment:
One testing pattern I have found value in is to just have a `List[String]` of all CLI invocations that are important to not regress, and just foreach assert that they don't result in a parse error. 

An example of this is if you use a command in cron, or something similar. This can help prevent it from being broken accidentally, and it's much lighter weight to just add the one invocation string to the "doesn't fail" list than a deep test. Often that's all you need

Reply:
Yeah, good call -- I've seen that pattern work well too. Added a quick mention, but may expand more as I edit...